keras_nlp/models/bloom/bloom_causal_lm.py

```diff
@@ -0,0 +1,319 @@
+
+        # === Default compilation ===
+        self.compile(
+            loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+            optimizer=keras.optimizers.Adam(2e-5),
+            metrics=[keras.metrics.SparseCategoricalAccuracy()],
+            sampler="greedy",
+            jit_compile=True,
+        )
+
+    @classproperty
+    def presets(cls):
+        return copy.deepcopy(backbone_presets)
+
+    @classproperty
+    def backbone_cls(cls):
+        return BloomBackbone
+
+    @classproperty
+    def preprocessor_cls(cls):
+        return BloomCausalLMPreprocessor
+
+    def call_with_cache(
+        self,
+        token_ids,
+        cache,
+        cache_update_index,
+    ):
+        """Forward pass of `BloomCausalLM` with cache.
+
+        `call_with_cache` adds an additional forward pass for the model for
+        autoregressive inference. Unlike calling the model directly, this method
+        allows caching previous key/value Tensors in multi-head attention layer,
+        and avoids recomputing the outputs of seen tokens.
+
+        Args:
+            token_ids: a dense int Tensor with shape `(batch_size, max_length)`.
+            cache: a dense float Tensor, the cache of key and value.
+            cache_update_index: int, or int Tensor. The index of current inputs
+                in the whole sequence.
+
+        Returns:
+            A (logits, hidden_states, cache) tuple. Where `logits` is the
+            language model logits for the input token_ids, `hidden_states` is
+            the final hidden representation of the input tokens, and `cache` is
+            the decoding cache.
+        """
+        x = self.backbone.token_embedding(token_ids)
+        x = self.backbone.embeddings_layer_norm(x)
+        # Each decoder layer has a cache; we update them separately.
+        caches = []
+        for i, transformer_layer in enumerate(self.backbone.transformer_layers):
+            current_cache = cache[:, i, ...]
+            x, next_cache = transformer_layer(
+                x,
+                cache=current_cache,
+                cache_update_index=cache_update_index,
+            )
+            caches.append(next_cache)
+        cache = ops.stack(caches, axis=1)
+        hidden_states = x = self.backbone.layer_norm(x)
+        logits = self.backbone.token_embedding(x, reverse=True)
+        return logits, hidden_states, cache
+
+    def _build_cache(self, token_ids):
+        """Build an empty cache for use with `call_with_cache()`."""
+        batch_size = ops.shape(token_ids)[0]
+        max_length = ops.shape(token_ids)[1]
+        num_layers = self.backbone.num_layers
+        num_heads = self.backbone.num_heads
+        head_dim = self.backbone.hidden_dim // num_heads
+        shape = [batch_size, num_layers, 2, max_length, num_heads, head_dim]
+        attention_layer_dtype = self.backbone.transformer_layers[0].dtype
+        cache = ops.zeros(shape, dtype=attention_layer_dtype)
```

Comment:
I added these two lines because of a keras2 error. Setting the cache dtype to the same dtype as the variable_dtype of the attention layer will prevent raising the tensorflow error while updating the cache().
```
TypeError: Input 'update' of 'XlaDynamicUpdateSlice' Op has type bfloat16 that does not match type float32 of argument 'input'.
```
This error didn't raise in keras3, because keras3 casts all the inputs dtype to the layer.variable_dtype,

Reply:
I actually think we want this to be `compute_dtype` not `variable_dtype`, for mixed precision cases. In that case, a variable dtype might be `float32`, but the output of layers will be `bfloat16` or `float16`. This cache is the cache our key/value projection outputs, so the compute dtype would be the correct thing to cache here.

Do you run into errors if you make that change?

Reply:
Thank you for that! That's right, the cache dtype should be the same as the dtype of the einsum layer output. That was my intention actually, but I thought that the output of the layer will be the `variable_dtype` not `compute_dtype`. 

But are you with me that this is not compatible with keras2:
```
        cache = ops.zeros(shape, dtype=self.compute_dtype)
```
because `BloomCausalLM` doesn't have the same dtype policy as the backbone layers, it follow the global policy. 
So instead we should initialize the cache with the compute_dtype of the backbone attention layers policy.
like that:
```
        attention_layer_dtype = self.backbone.transformer_layers[
            0
        ].compute_dtype
        cache = ops.zeros(shape, dtype=attention_layer_dtype)
```
Even we can't do:
```
        cache = ops.zeros(shape, dtype=self.backbone.compute_dtype)
```
because also the backbone has dtype policy different than its layers ( when we pass a `dtype` arg different from the global policy during initialization).

Take a look at this commit :
https://github.com/keras-team/keras-nlp/pull/1467/commits/77c576bf2002d4c3889bcc83444b870edcad3e7e
to show how we can set the dtype_policy of the backbone. and how to set the causalLM to follow the backbone dtype_policy. I will revert this commit after checking the tests because it's not its time now. but you can take a look. I can open a PR to Apply that for all models, if you think it will be useful.

Reply:
Also I forget to mention, Look at that commit and its error:
https://github.com/keras-team/keras-nlp/pull/1467/commits/7cc8671097277c3c2503c0fe6d98facabe87ecf6
That will clarify what I am talking about a little bit

Reply:
Not a computer right now, but I think you are right. We want to set the task dtype policy to just follow the backbone's exactly. And set the backbone dtype accessors so they reflect the proper dtype. That's a bug we can probably fix on a separate pr

Reply:
Will land https://github.com/keras-team/keras-nlp/pull/1486 tomorrow, and pull this in with `self.compute_dtype`.

Looks great!

Reply:
Ok #1486 is in. Shall I update this to `self.compute_dtype` or do you want to?